pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -335,102 +352,253 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+			else
+			  -[#orangered]-> yes;
+			  #orangered:<<implicit>>\n**Available = False** /
+			endif
+		  endif
+		endif
+		end
+		@enduml
+	*/
+
+	// If any component operator reports negatively we want to write that to
+	// the instance while preserving it's lastTransitionTime.
+	// For example, consider the KubeVirt resource has the Available condition
+	// type with type "False". When reconciling KubeVirt's resource we would
+	// add it to the in-memory representation of HCO's conditions (r.conditions)
+	// and here we are simply writing it back to the server.
+	// One shortcoming is that only one failure of a particular condition can be
+	// captured at one time (ie. if KubeVirt and CDI are both reporting !Available,
+	// you will only see CDI as it updates last).
+
+	req.conditions.setStatusCondition(conditionsv1.Condition{
 		Type:    hcov1alpha1.ConditionReconcileComplete,
 		Status:  corev1.ConditionTrue,
 		Reason:  reconcileCompleted,
 		Message: reconcileCompletedMessage,
 	})
 
-	// Requeue if we just created everything
-	if init {
-		return reconcile.Result{Requeue: true}, nil
-	}
+	if cond, conditionFound := req.conditions[conditionsv1.ConditionDegraded]; conditionFound { // (#chart 1)
+		if _, conditionFound = req.conditions[conditionsv1.ConditionProgressing]; !conditionFound { // (#chart 2)
+			req.conditions.setStatusCondition(conditionsv1.Condition{ // (#chart 3)
+				Type:    conditionsv1.ConditionProgressing,
+				Status:  corev1.ConditionFalse,
+				Reason:  reconcileCompleted,
+				Message: reconcileCompletedMessage,
+			})
+		} // else - Progressing is already exists
 
-	if r.conditions == nil {
-		reqLogger.Info("No component operator reported negatively")
-		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
-			Type:    conditionsv1.ConditionAvailable,
-			Status:  corev1.ConditionTrue,
-			Reason:  reconcileCompleted,
-			Message: reconcileCompletedMessage,
-		})
-		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
-			Type:    conditionsv1.ConditionProgressing,
-			Status:  corev1.ConditionFalse,
-			Reason:  reconcileCompleted,
-			Message: reconcileCompletedMessage,
-		})
-		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+		if _, conditionFound = req.conditions[conditionsv1.ConditionUpgradeable]; !conditionFound { // (#chart 4)
+			req.conditions.setStatusCondition(conditionsv1.Condition{ // (#chart 5)
+				Type:    conditionsv1.ConditionUpgradeable,
+				Status:  corev1.ConditionFalse,
+				Reason:  cond.Reason,
+				Message: cond.Reason,
+			})
+		} // else - Upgradeable is already exists
+		if _, conditionFound = req.conditions[conditionsv1.ConditionAvailable]; !conditionFound { // (#chart 6)
+			req.conditions.setStatusCondition(conditionsv1.Condition{ // (#chart 7)
+				Type:    conditionsv1.ConditionAvailable,
+				Status:  corev1.ConditionFalse,
+				Reason:  cond.Reason,
+				Message: cond.Reason,
```

Comment:
Still not 100% confident about this:
basically if we found a component that reported degraded we are also simply duplicating the error message we got from that component also on available condition on HCO CR.
So when the user tries to check available condition on HCO CR he will find something like "XXX is degraded due to..." that can sound a bit misleading.
Maybe in this cases we can simply use a placeholder message explicitly saying that HCO is not available because we have degraded components.